vpr/src/analytical_place/analytical_solver.cpp

```diff
@@ -0,0 +1,264 @@
+        VTR_ASSERT_DEBUG(src_row_id < A_sparse.rows());
+        // Verify that this is a valid block id.
+        VTR_ASSERT_DEBUG(target_blk_id.is_valid());
+        // The src_row_id is always a moveable block (rows in the matrix always
+        // coorespond to a moveable APBlock or a star node.
+        if (netlist_.block_mobility(target_blk_id) == APBlockMobility::MOVEABLE) {
+            // If the target is also moveable, update the coefficient matrix.
+            size_t target_row_id = (size_t)blk_id_to_row_id_[target_blk_id];
+            VTR_ASSERT_DEBUG(target_row_id < A_sparse.rows());
+            tripletList.emplace_back(src_row_id, src_row_id, weight);
+            tripletList.emplace_back(target_row_id, target_row_id, weight);
+            tripletList.emplace_back(src_row_id, target_row_id, -weight);
+            tripletList.emplace_back(target_row_id, src_row_id, -weight);
+        } else {
+            // If the target is fixed, update the coefficient matrix and the
+            // constant vectors.
+            tripletList.emplace_back(src_row_id, src_row_id, weight);
+            VTR_ASSERT_DEBUG(netlist_.block_loc(target_blk_id).x >= 0);
+            VTR_ASSERT_DEBUG(netlist_.block_loc(target_blk_id).y >= 0);
+            // FIXME: These fixed block locations are aligned to the anchor of
+            //        the tiles they are in. This is not correct. A method
+            //        should be added to the netlist class or to a util file
+            //        which can get a more accurate position.
+            double blk_loc_x = netlist_.block_loc(target_blk_id).x;
+            double blk_loc_y = netlist_.block_loc(target_blk_id).y;
+            b_x(src_row_id) += weight * blk_loc_x;
+            b_y(src_row_id) += weight * blk_loc_y;
+        }
+    };
+
+    // Create the connections using a hybrid connection model of the star and
+    // clique connnection models.
+    size_t star_node_offset = 0;
+    for (APNetId net_id : netlist_.nets()) {
+        size_t num_pins = netlist_.net_pins(net_id).size();
+        VTR_ASSERT_DEBUG(num_pins > 1);
+        if (num_pins > star_num_pins_threshold) {
+            // Create a star node and connect each block in the net to the star
+            // node.
+            // Using the weight from FastPlace
+            double w = static_cast<double>(num_pins) / static_cast<double>(num_pins - 1);
+            size_t star_node_id = num_moveable_blocks_ + star_node_offset;
+            for (APPinId pin_id : netlist_.net_pins(net_id)) {
+                APBlockId blk_id = netlist_.pin_block(pin_id);
+                add_connection_to_system(star_node_id, blk_id, w);
+            }
+            star_node_offset++;
+        } else {
+            // Create a clique connection where every block in a net connects
+            // exactly once to every other block in the net.
+            // Using the weight from FastPlace
+            double w = 1.0 / static_cast<double>(num_pins - 1);
```

Comment:
This might give too much weight to high fanout nets (n^2 edges * 1/n weight --> n weight for an n-pin net).  ASIC placers may run with more controlled fanout (generally I'd expect some buffer insertion was done before running an ASIC placer) so the scaling of this to the very high fanouts in FPGA nets may not work out well.  I'd add a note / TODO for that.

Reply:
This is the star node weight. For a net with P pins, there will be P edges created or the star net model. This weight factor is a correction term to account for these P extra pins. Hence why it is a linear correct. This is from the FastPlace paper. We can explore other weight terms, but I am not sure if they approach the correct solution. I will add a TODO to investigate.